File: SDF.CSharp/MeshGenerator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Numerics;
using System.Threading.Tasks;

namespace SDF;

/// <summary>
/// Core mesh generation engine
/// </summary>
public class MeshGenerator
{
    public int Workers { get; set; } = Environment.ProcessorCount;
    public int Samples { get; set; } = 1 << 22; // 2^22
    public int BatchSize { get; set; } = 32;
    public bool Sparse { get; set; } = true;
    public bool Verbose { get; set; } = true;

    /// <summary>
    /// Generate a mesh from an SDF
    /// </summary>
    public List<Vector3> Generate(SDF3 sdf, 
        double? step = null, 
        (Vector3, Vector3)? bounds = null)
    {
        var (min, max) = bounds ?? EstimateBounds(sdf);
        
        if (Verbose)
        {
            Console.WriteLine($"Bounds: ({min.X:F2}, {min.Y:F2}, {min.Z:F2}) to ({max.X:F2}, {max.Y:F2}, {max.Z:F2})");
        }

        // Calculate step size if not provided
        if (!step.HasValue)
        {
            var volume = (max.X - min.X) * (max.Y - min.Y) * (max.Z - min.Z);
            step = Math.Pow(volume / Samples, 1.0 / 3.0);
        }

        if (Verbose)
        {
            Console.WriteLine($"Step size: {step:F6}");
        }

        // Calculate grid dimensions
        var nx = (int)Math.Ceiling((max.X - min.X) / step.Value);
        var ny = (int)Math.Ceiling((max.Y - min.Y) / step.Value);
        var nz = (int)Math.Ceiling((max.Z - min.Z) / step.Value);

        if (Verbose)
        {
            Console.WriteLine($"Grid dimensions: {nx} x {ny} x {nz}");
            Console.WriteLine("Generating mesh...");
        }

        // Create batches
        var batches = CreateBatches(min, max, step.Value, nx, ny, nz);
        
        if (Verbose)
        {
            Console.WriteLine($"Processing {batches.Count} batches...");
        }

        // Process batches in parallel
        var allTriangles = new List<Vector3>();
        var lockObj = new object();
        
        Parallel.ForEach(batches, new ParallelOptions { MaxDegreeOfParallelism = Workers }, batch =>
        {
            var triangles = ProcessBatch(sdf, batch);
            if (triangles != null && triangles.Count > 0)
            {
                lock (lockObj)
                {
                    allTriangles.AddRange(triangles);
                }
            }
        });

        if (Verbose)
        {
            Console.WriteLine($"Generated {allTriangles.Count / 3} triangles");
        }

        return allTriangles;
    }

    private List<Vector3>? ProcessBatch(SDF3 sdf, Batch batch)
    {
        // Skip batch if it's far from the surface (for sparse sampling)
        if (Sparse && ShouldSkip(sdf, batch))
        {
            return null;
        }

        // Sample the SDF at all grid points in the batch
        var points = new List<Vector3>();
        for (int x = 0; x < batch.Nx; x++)
        {
            for (int y = 0; y < batch.Ny; y++)
            {
                for (int z = 0; z < batch.Nz; z++)
                {
                    points.Add(new Vector3(
                        batch.MinX + x * batch.StepX,
                        batch.MinY + y * batch.StepY,
                        batch.MinZ + z * batch.StepZ
                    ));
                }
            }
        }

        var values = sdf.Evaluate(points.ToArray());
        
        // Reshape into 3D array
        var volume = new double[batch.Nx, batch.Ny, batch.Nz];
        int idx = 0;
        for (int x = 0; x < batch.Nx; x++)
        {
            for (int y = 0; y < batch.Ny; y++)
            {
                for (int z = 0; z < batch.Nz; z++)
                {
                    volume[x, y, z] = values[idx++];
                }
            }
        }

        // Apply marching cubes
        var step = new Vector3(batch.StepX, batch.StepY, batch.StepZ);
        var offset = new Vector3(batch.MinX, batch.MinY, batch.MinZ);
        return MarchingCubes.Generate(volume, offset, step);
    }

    private bool ShouldSkip(SDF3 sdf, Batch batch)
    {
        // Check center point and corners
        var center = new Vector3(
            (batch.MinX + batch.MaxX) / 2,
            (batch.MinY + batch.MaxY) / 2,
            (batch.MinZ + batch.MaxZ) / 2
        );

        var r = Math.Abs(sdf.Evaluate(new[] { center })[0]);
        var dx = batch.MaxX - batch.MinX;
        var dy = batch.MaxY - batch.MinY;
        var dz = batch.MaxZ - batch.MinZ;
        var d = Math.Sqrt(dx * dx + dy * dy + dz * dz) / 2;

        if (r > d)
        {
            // Check corners
            var corners = new[]
            {
                new Vector3(batch.MinX, batch.MinY, batch.MinZ),
                new Vector3(batch.MaxX, batch.MinY, batch.MinZ),
                new Vector3(batch.MinX, batch.MaxY, batch.MinZ),
                new Vector3(batch.MaxX, batch.MaxY, batch.MinZ),
                new Vector3(batch.MinX, batch.MinY, batch.MaxZ),
                new Vector3(batch.MaxX, batch.MinY, batch.MaxZ),
                new Vector3(batch.MinX, batch.MaxY, batch.MaxZ),
                new Vector3(batch.MaxX, batch.MaxY, batch.MaxZ),
            };

            var cornerValues = sdf.Evaluate(corners);
            var allPositive = cornerValues.All(v => v > 0);
            var allNegative = cornerValues.All(v => v < 0);
            
            return allPositive || allNegative;
        }

        return false;
    }

    private List<Batch> CreateBatches(Vector3 min, Vector3 max, double step, int nx, int ny, int nz)
    {
        var batches = new List<Batch>();
        var batchSize = BatchSize;

        for (int bx = 0; bx < nx; bx += batchSize)
        {
            for (int by = 0; by < ny; by += batchSize)
            {
                for (int bz = 0; bz < nz; bz += batchSize)
                {
                    var batch = new Batch
                    {
                        MinX = (float)(min.X + bx * step),
                        MinY = (float)(min.Y + by * step),
                        MinZ = (float)(min.Z + bz * step),
                        Nx = Math.Min(batchSize + 1, nx - bx + 1),
                        Ny = Math.Min(batchSize + 1, ny - by + 1),
                        Nz = Math.Min(batchSize + 1, nz - bz + 1),
                        StepX = (float)step,
                        StepY = (float)step,
                        StepZ = (float)step
                    };
                    batch.MaxX = (float)(batch.MinX + (batch.Nx - 1) * step);
                    batch.MaxY = (float)(batch.MinY + (batch.Ny - 1) * step);
                    batch.MaxZ = (float)(batch.MinZ + (batch.Nz - 1) * step);
                    
                    batches.Add(batch);
                }
            }
        }

        return batches;
    }

    public (Vector3, Vector3) EstimateBounds(SDF3 sdf)
    {
        const int samples = 16;
        var min = new Vector3(-1e9f, -1e9f, -1e9f);
        var max = new Vector3(1e9f, 1e9f, 1e9f);
        double prevThreshold = -1;

        for (int iteration = 0; iteration < 32; iteration++)
        {
            var dx = (max.X - min.X) / (samples - 1);
            var dy = (max.Y - min.Y) / (samples - 1);
            var dz = (max.Z - min.Z) / (samples - 1);
            
            var threshold = Math.Sqrt(dx * dx + dy * dy + dz * dz) / 2;
            
            if (Math.Abs(threshold - prevThreshold) < 1e-10)
                break;
            
            prevThreshold = threshold;

            var points = new List<Vector3>();
            for (int x = 0; x < samples; x++)
            {
                for (int y = 0; y < samples; y++)
                {
                    for (int z = 0; z < samples; z++)
                    {
                        points.Add(new Vector3(
                            min.X + x * dx,
                            min.Y + y * dy,
                            min.Z + z * dz
                        ));
                    }
                }
            }

            var values = sdf.Evaluate(points.ToArray());
            var nearSurface = new List<Vector3>();
            
            for (int i = 0; i < points.Count; i++)
            {
                if (Math.Abs(values[i]) <= threshold)
                {
                    nearSurface.Add(points[i]);
                }
            }

            if (nearSurface.Count == 0)
                break;

            var newMin = nearSurface[0];
            var newMax = nearSurface[0];
            
            foreach (var p in nearSurface)
            {
                newMin = new Vector3(
                    Math.Min(newMin.X, p.X),
                    Math.Min(newMin.Y, p.Y),
                    Math.Min(newMin.Z, p.Z)
                );
                newMax = new Vector3(
                    Math.Max(newMax.X, p.X),
                    Math.Max(newMax.Y, p.Y),
                    Math.Max(newMax.Z, p.Z)
                );
            }

            min = newMin - new Vector3(dx, dy, dz) / 2;
            max = newMax + new Vector3(dx, dy, dz) / 2;
        }

        return (min, max);
    }

    private class Batch
    {
        public float MinX, MinY, MinZ;
        public float MaxX, MaxY, MaxZ;
        public int Nx, Ny, Nz;
        public float StepX, StepY, StepZ;
    }
}

File: SDF.Cpp/src/MeshGenerator.cpp
#include "sdf/MeshGenerator.h"
#include "sdf/MarchingCubes.h"
#include <cmath>
#include <algorithm>
#include <thread>
#include <mutex>
#include <atomic>
#include <iostream>
#include <chrono>

namespace sdf {

// Helper for clamp (C++17 compatible)
template<typename T>
static T clamp(T value, T min, T max) {
    return std::max(min, std::min(value, max));
}

std::vector<Vector3> MeshGenerator::generate(const SDF3& sdf, const Options& options) {
    auto startTime = std::chrono::high_resolution_clock::now();
    
    // Determine bounds
    Vector3 boundsMin, boundsMax;
    if (options.customBounds) {
        boundsMin = options.boundsMin;
        boundsMax = options.boundsMax;
    } else {
        auto bounds = estimateBounds(sdf);
        boundsMin = bounds.first;
        boundsMax = bounds.second;
    }
    
    if (options.verbose) {
        std::cout << "Bounds: (" << boundsMin << ") to (" << boundsMax << ")\n";
    }
    
    // Calculate step size
    double step = options.step;
    if (step == 0.0 && options.samples > 0) {
        double volume = (boundsMax.x - boundsMin.x) * 
                       (boundsMax.y - boundsMin.y) * 
                       (boundsMax.z - boundsMin.z);
        step = std::cbrt(volume / options.samples);
    }
    if (step == 0.0) {
        step = 0.1;  // Default step
    }
    
    // Calculate grid dimensions
    int nx = static_cast<int>(std::ceil((boundsMax.x - boundsMin.x) / step)) + 1;
    int ny = static_cast<int>(std::ceil((boundsMax.y - boundsMin.y) / step)) + 1;
    int nz = static_cast<int>(std::ceil((boundsMax.z - boundsMin.z) / step)) + 1;
    
    if (options.verbose) {
        std::cout << "Grid dimensions: " << nx << " x " << ny << " x " << nz << "\n";
        std::cout << "Step size: " << step << "\n";
    }
    
    // Adjust bounds to match grid
    boundsMax.x = boundsMin.x + (nx - 1) * step;
    boundsMax.y = boundsMin.y + (ny - 1) * step;
    boundsMax.z = boundsMin.z + (nz - 1) * step;
    
    // Create batches
    int batchSize = options.batchSize;
    std::vector<std::tuple<int, int, int, int, int, int>> batches;
    
    for (int z = 0; z < nz - 1; z += batchSize) {
        for (int y = 0; y < ny - 1; y += batchSize) {
            for (int x = 0; x < nx - 1; x += batchSize) {
                int x1 = std::min(x + batchSize, nx - 1);
                int y1 = std::min(y + batchSize, ny - 1);
                int z1 = std::min(z + batchSize, nz - 1);
                batches.push_back(std::make_tuple(x, x1, y, y1, z, z1));
            }
        }
    }
    
    if (options.verbose) {
        std::cout << "Processing " << batches.size() << " batches...\n";
    }
    
    // Determine worker count
    int workers = options.workers;
    if (workers == 0) {
        workers = std::thread::hardware_concurrency();
        if (workers == 0) workers = 4;
    }
    
    // Process batches in parallel
    std::vector<Vector3> allTriangles;
    std::mutex resultMutex;
    std::atomic<int> batchesProcessed(0);
    std::atomic<int> batchesSkipped(0);
    
    auto processBatchRange = [&](int start, int end) {
        std::vector<Vector3> localTriangles;
        
        for (int i = start; i < end; ++i) {
            auto [x0, x1, y0, y1, z0, z1] = batches[i];
            
            Vector3 batchMin(
                boundsMin.x + x0 * step,
                boundsMin.y + y0 * step,
                boundsMin.z + z0 * step
            );
            Vector3 batchMax(
                boundsMin.x + x1 * step,
                boundsMin.y + y1 * step,
                boundsMin.z + z1 * step
            );
            
            // Skip batch if far from surface (sparse sampling)
            if (options.sparse && canSkipBatch(sdf, batchMin, batchMax)) {
                batchesSkipped++;
                batchesProcessed++;
                continue;
            }
            
            // Generate grid points for this batch
            int batchNx = x1 - x0 + 1;
            int batchNy = y1 - y0 + 1;
            int batchNz = z1 - z0 + 1;
            
            std::vector<Vector3> gridPoints;
            gridPoints.reserve(batchNx * batchNy * batchNz);
            
            for (int bz = 0; bz < batchNz; ++bz) {
                for (int by = 0; by < batchNy; ++by) {
                    for (int bx = 0; bx < batchNx; ++bx) {
                        gridPoints.push_back(Vector3(
                            boundsMin.x + (x0 + bx) * step,
                            boundsMin.y + (y0 + by) * step,
                            boundsMin.z + (z0 + bz) * step
                        ));
                    }
                }
            }
            
            // Evaluate SDF at all grid points
            auto values = sdf.evaluate(gridPoints);
            
            // Run marching cubes
            std::array<int, 3> dims = {batchNx, batchNy, batchNz};
            auto batchTriangles = MarchingCubes::extractSurface(values, dims, 0.0);
            
            // Scale and translate triangles
            Vector3 scale(step, step, step);
            Vector3 offset = batchMin;
            
            for (auto& v : batchTriangles) {
                v = v * scale + offset;
            }
            
            localTriangles.insert(localTriangles.end(), 
                                 batchTriangles.begin(), 
                                 batchTriangles.end());
            
            batchesProcessed++;
        }
        
        // Merge results
        std::lock_guard<std::mutex> lock(resultMutex);
        allTriangles.insert(allTriangles.end(), 
                           localTriangles.begin(), 
                           localTriangles.end());
    };
    
    // Launch threads
    std::vector<std::thread> threads;
    int batchesPerWorker = (batches.size() + workers - 1) / workers;
    
    for (int i = 0; i < workers; ++i) {
        int start = i * batchesPerWorker;
        int end = std::min(start + batchesPerWorker, static_cast<int>(batches.size()));
        if (start < end) {
            threads.emplace_back(processBatchRange, start, end);
        }
    }
    
    // Wait for completion
    for (auto& thread : threads) {
        thread.join();
    }
    
    auto endTime = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(endTime - startTime);
    
    if (options.verbose) {
        std::cout << "Generated " << (allTriangles.size() / 3) << " triangles\n";
        std::cout << "Processed " << batchesProcessed << " batches (skipped " 
                  << batchesSkipped << ")\n";
        std::cout << "Time: " << (duration.count() / 1000.0) << " seconds\n";
    }
    
    return allTriangles;
}

std::pair<Vector3, Vector3> MeshGenerator::estimateBounds(const SDF3& sdf) {
    const int samples = 16;
    Vector3 boundsMin(-10, -10, -10);
    Vector3 boundsMax(10, 10, 10);
    
    double prevThreshold = -1.0;
    
    for (int iteration = 0; iteration < 32; ++iteration) {
        Vector3 step = (boundsMax - boundsMin) / (samples - 1);
        double threshold = step.length() / 2.0;
        
        if (std::abs(threshold - prevThreshold) < 1e-10) {
            break;
        }
        prevThreshold = threshold;
        
        // Sample points on a grid
        std::vector<Vector3> points;
        points.reserve(samples * samples * samples);
        
        for (int z = 0; z < samples; ++z) {
            for (int y = 0; y < samples; ++y) {
                for (int x = 0; x < samples; ++x) {
                    points.push_back(Vector3(
                        boundsMin.x + x * step.x,
                        boundsMin.y + y * step.y,
                        boundsMin.z + z * step.z
                    ));
                }
            }
        }
        
        // Evaluate SDF
        auto values = sdf.evaluate(points);
        
        // Find points near surface
        Vector3 newMin = boundsMax;
        Vector3 newMax = boundsMin;
        bool foundAny = false;
        
        for (size_t i = 0; i < points.size(); ++i) {
            if (std::abs(values[i]) <= threshold) {
                newMin = Vector3::min(newMin, points[i]);
                newMax = Vector3::max(newMax, points[i]);
                foundAny = true;
            }
        }
        
        if (!foundAny) {
            // No surface found, expand bounds
            Vector3 center = (boundsMin + boundsMax) * 0.5;
            Vector3 size = boundsMax - boundsMin;
            boundsMin = center - size;
            boundsMax = center + size;
        } else {
            // Add margin
            Vector3 margin = step * 0.5;
            boundsMin = newMin - margin;
            boundsMax = newMax + margin;
        }
    }
    
    return {boundsMin, boundsMax};
}

bool MeshGenerator::canSkipBatch(
    const SDF3& sdf,
    const Vector3& batchMin,
    const Vector3& batchMax
) {
    // Sample center and corners
    Vector3 center = (batchMin + batchMax) * 0.5;
    double radius = (batchMax - batchMin).length() / 2.0;
    
    // Check center distance
    double centerDist = sdf(center);
    if (std::abs(centerDist) <= radius) {
        return false;  // Surface might be in batch
    }
    
    // Check all 8 corners
    std::vector<Vector3> corners = {
        Vector3(batchMin.x, batchMin.y, batchMin.z),
        Vector3(batchMax.x, batchMin.y, batchMin.z),
        Vector3(batchMin.x, batchMax.y, batchMin.z),
        Vector3(batchMax.x, batchMax.y, batchMin.z),
        Vector3(batchMin.x, batchMin.y, batchMax.z),
        Vector3(batchMax.x, batchMin.y, batchMax.z),
        Vector3(batchMin.x, batchMax.y, batchMax.z),
        Vector3(batchMax.x, batchMax.y, batchMax.z)
    };
    
    auto values = sdf.evaluate(corners);
    
    // Check if all corners have the same sign
    bool allPositive = true;
    bool allNegative = true;
    for (double v : values) {
        if (v < 0) allPositive = false;
        if (v > 0) allNegative = false;
    }
    
    return allPositive || allNegative;
}

} // namespace sdf
